### ISSUE TITLE
JET.jl reports possible errors on `AD.gradient` that do not appear with `Zygote.gradient`

### ISSUE THREAD

**Issue by gdalle** [JuliaDiff/AbstractDifferentiation.jl#55]
I recently started to use [JET.jl](https://github.com/aviatesk/JET.jl) for CI, and it turns out that some errors are reported that do not exist when using Zygote.jl directly.

MWE (short version, the full JET output is in the txt file I attached):

```julia
julia> using AbstractDifferentiation, JET, Zygote

julia> f(x) = sum(abs2, x)
f (generic function with 1 method)

julia> x = rand(100);

julia> ab = AD.ZygoteBackend()
AbstractDifferentiation.ReverseRuleConfigBackend{Zygote.ZygoteRuleConfig{Zygote.Context}}(Zygote.ZygoteRuleConfig{Zygote.Context}(Zygote.Context(nothing)))

julia> @report_call vscode_console_output=stdout Zygote.gradient(f, x)
No errors detected

julia> @report_call vscode_console_output=stdout AD.gradient(ab, f, x)
═════ 31 possible errors found ═════
```

[JET_report.txt](https://github.com/JuliaDiff/AbstractDifferentiation.jl/files/8354913/JET_report.txt)

**Comment by gdalle** [comment]
[JET_report.txt](https://github.com/JuliaDiff/AbstractDifferentiation.jl/files/8354912/JET_report.txt)

**Comment by ToucheSir** [comment]
Have you tested `rrule_via_ad(ZygoteRuleConfig(), ...)`. I would bet AD.jl is not at fault here.

**Comment by gdalle** [comment]
Doesn't seem to be the cause:
```julia
julia> using ChainRulesCore

julia> rc = Zygote.ZygoteRuleConfig();

julia> @report_call vscode_console_output=stdout rrule_via_ad(rc, f, x)
No errors detected

julia> _, pb = ChainRulesCore.rrule_via_ad(rc, f, x);

julia> @report_call vscode_console_output=stdout pb(1.)
No errors detected
```

**Comment by ToucheSir** [comment]
Since this was posted I discovered that AD.jl tries to calculate the full jacobian when you ask for gradients with the Zygote backend: https://github.com/JuliaDiff/AbstractDifferentiation.jl/issues/63#issuecomment-1225959150. That's likely the reason for the type instability...

**Comment by devmotion** [comment]
Fixed by #128. On the master branch I get
```julia
ulia> @report_call vscode_console_output=stdout Zygote.gradient(f, x)
═════ 1 possible error found ═════
┌ gradient(f::typeof(f), args::Vector{Float64}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/interface.jl:149
│┌ _project_all(x::Tuple{Vector{Float64}}, dx::Tuple{Vector{Float64}}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/interface.jl:118
││┌ map(f::typeof(Zygote._project), t::Tuple{Vector{Float64}}, s::Tuple{Vector{Float64}}) @ Base ./tuple.jl:318
│││┌ _project(x::Vector{Float64}, dx::Vector{Float64}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/chainrules.jl:189
││││┌ (::ChainRulesCore.ProjectTo{AbstractArray, @NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}})(dx::Vector{Float64}) @ ChainRulesCore /Users/david/.julia/packages/ChainRulesCore/PvTbU/src/projection.jl:238
│││││┌ getproperty(p::ChainRulesCore.ProjectTo{AbstractArray, @NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}}, name::Symbol) @ ChainRulesCore /Users/david/.julia/packages/ChainRulesCore/PvTbU/src/projection.jl:38
││││││┌ getproperty(x::@NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}, f::Symbol) @ Base ./Base.jl:37
│││││││ type NamedTuple has no field elements: Base.getfield(x::@NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}, f::Symbol)
││││││└────────────────────


julia> @report_call vscode_console_output=stdout AD.gradient(ab, f, x)
═════ 1 possible error found ═════
┌ gradient(::AbstractDifferentiation.ZygoteBackend, f::typeof(f), args::Vector{Float64}) @ AbstractDifferentiationZygoteExt /Users/david/.julia/packages/AbstractDifferentiation/jhFqr/ext/AbstractDifferentiationZygoteExt.jl:20
│┌ gradient(f::typeof(f), args::Vector{Float64}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/interface.jl:149
││┌ _project_all(x::Tuple{Vector{Float64}}, dx::Tuple{Vector{Float64}}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/interface.jl:118
│││┌ map(f::typeof(Zygote._project), t::Tuple{Vector{Float64}}, s::Tuple{Vector{Float64}}) @ Base ./tuple.jl:318
││││┌ _project(x::Vector{Float64}, dx::Vector{Float64}) @ Zygote /Users/david/.julia/packages/Zygote/jxHJc/src/compiler/chainrules.jl:189
│││││┌ (::ChainRulesCore.ProjectTo{AbstractArray, @NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}})(dx::Vector{Float64}) @ ChainRulesCore /Users/david/.julia/packages/ChainRulesCore/PvTbU/src/projection.jl:238
││││││┌ getproperty(p::ChainRulesCore.ProjectTo{AbstractArray, @NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}}, name::Symbol) @ ChainRulesCore /Users/david/.julia/packages/ChainRulesCore/PvTbU/src/projection.jl:38
│││││││┌ getproperty(x::@NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}, f::Symbol) @ Base ./Base.jl:37
││││││││ type NamedTuple has no field elements: Base.getfield(x::@NamedTuple{element::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}}, axes::Tuple{Base.OneTo{Int64}}}, f::Symbol)
│││││││└────────────────────
```
